lib/functions/versions.bash

```diff
@@ -161,36 +161,72 @@ latest_all() {
     for plugin_path in "$plugins_path"/*/; do
       plugin_name=$(basename "$plugin_path")
 
+      local installed_versions
+      installed_versions=$(list_installed_versions "$plugin_name")
+
+      # Get a distinct list of distributions/variants
+      local installed_variants
+      installed_variants=$(echo "$installed_versions" |
+        sed -e "s/\(^.*\)\-.*/\1/" |        # extract variant prefix
+        sort -u |                           # unique
+        sed -e "s/^[[:space:]]*[0-9].*//" | # remove versions without variant prefix
+        sed '/^[[:space:]]*$/d')            # remove empty lines
+
       # Retrieve the version of the plugin
-      local version
+      local versions=()
       if [ -f "${plugin_path}/bin/latest-stable" ]; then
         # We can't filter by a concrete query because different plugins might
         # have different queries.
+        local version
         version=$("${plugin_path}"/bin/latest-stable "")
+
         if [ -z "${version}" ]; then
-          version="unknown"
+          versions+=("unknown")
+        else
+          versions+=("$version")
         fi
+      elif [ -n "$installed_variants" ]; then
+        # get latest version for each variant
+        for variant in $installed_variants; do
+          local version
+          version=$(list_all_command "$plugin_name" "$(echo "$variant" | tr -d '[:space:]')" |
+            grep -ivE "(^Available version:|-src|-dev|-latest|-stm|[-\\.]rc|-alpha|-beta|[-\\.]pre|-next|(a|b|c)[0-9]+|snapshot|master)" |
+            # filter for exact match or with version suffix to deal with overlaps (e.g. zulu-18.28.13, zulu-jre-18.28.13)
+            grep -iE "^($variant|$variant(-\d+.*)?)\$" |
+            sed 's/^[[:space:]]\+//' |
+            tail -1)
+
```

Comment:
While shorter to write this way, we're calling `list_all_command` for each variant when really it only needs to be called once for the plugin. The `list_all_command` executes `<plugin>/bin/list-all` which performs a network call and file read/writes. It is usually a really slow command. Can you rework this some more?

Reply:
Very Vvalid concern, will rework this part 👍